miner/scroll_worker.go

```diff
@@ -316,133 +352,152 @@ func (w *worker) mainLoop() {
+	if w.chainConfig.Clique != nil && w.chainConfig.Clique.RelaxedPeriod {
+		// clique with relaxed period uses time.Now() as the header.Time, calculate the deadline
+		deadline = time.Unix(int64(header.Time+w.chainConfig.Clique.Period), 0)
+	}
+
+	w.current = &work{
+		deadlineTimer:  time.NewTimer(time.Until(deadline)),
+		cccLogger:      cccLogger,
+		vmConfig:       vmConfig,
+		header:         header,
+		state:          parentState,
+		txs:            types.Transactions{},
+		receipts:       types.Receipts{},
+		coalescedLogs:  []*types.Log{},
+		gasPool:        new(core.GasPool).AddGas(header.GasLimit),
+		nextL1MsgIndex: nextL1MsgIndex,
+		reorgReason:    reorgReason,
+	}
+	return nil
+}
+
+// tryCommitNewWork
+func (w *worker) tryCommitNewWork(now time.Time, parent common.Hash, reorgReason error) (common.Hash, error) {
+	err := w.newWork(now, parent, reorgReason)
 	if err != nil {
-		log.Error("failed to fetch parent state", "err", err)
-		return
+		return common.Hash{}, fmt.Errorf("failed creating new work: %w", err)
 	}
 
-	// Apply special state transition at Curie block
-	if w.chainConfig.CurieBlock != nil && w.chainConfig.CurieBlock.Cmp(header.Number) == 0 {
-		misc.ApplyCurieHardFork(parentState)
+	shouldCommit, err := w.handleForks()
+	if err != nil {
+		return common.Hash{}, fmt.Errorf("failed handling forks: %w", err)
+	}
 
-		var nextL1MsgIndex uint64
-		if dbVal := rawdb.ReadFirstQueueIndexNotInL2Block(w.eth.ChainDb(), header.ParentHash); dbVal != nil {
-			nextL1MsgIndex = *dbVal
-		}
+	// check if we are reorging
+	reorging := w.chain.GetBlockByNumber(w.current.header.Number.Uint64()) != nil
+	if !shouldCommit && reorging {
+		shouldCommit, err = w.processReorgedTxns(w.current.reorgReason)
+	}
+	if err != nil {
+		return common.Hash{}, fmt.Errorf("failed handling reorged txns: %w", err)
+	}
 
-		// zkEVM requirement: Curie transition block contains 0 transactions, bypass pipeline.
-		err = w.commit(&pipeline.Result{
-			// Note: Signer nodes will not store CCC results for empty blocks in their database.
-			// In practice, this is acceptable, since this block will never overflow, and follower
-			// nodes will still store CCC results.
-			Rows: &types.RowConsumption{},
-			FinalBlock: &pipeline.BlockCandidate{
-				Header:         header,
-				State:          parentState,
-				Txs:            types.Transactions{},
-				Receipts:       types.Receipts{},
-				CoalescedLogs:  []*types.Log{},
-				NextL1MsgIndex: nextL1MsgIndex,
-			},
-		})
+	if !shouldCommit {
+		shouldCommit, err = w.processTxPool()
+	}
+	if err != nil {
+		return common.Hash{}, fmt.Errorf("failed processing tx pool: %w", err)
+	}
 
-		if err != nil {
-			log.Error("failed to commit Curie fork block", "reason", err)
+	if shouldCommit {
+		// if reorging, force committing even if we are not "running"
+		// this can happen when sequencer is instructed to shutdown while handling a reorg
+		// we should make sure reorg is not interrupted
```

Comment:
What happens if we're interrupted?

Even if that happens, we should detect that CCC was not run/stored on the last few blocks, and re-run after restart

Reply:
> What happens if we're interrupted?
Nothing catastrophic, it is just a more graceful way of shutting down.

> Even if that happens, we should detect that CCC was not run/stored on the last few blocks, and re-run after restart

Yeap, this was added in https://github.com/scroll-tech/go-ethereum/pull/982/commits/057b2c10c3738b9662fd68f4c86c88c5e581a233